internal/cli/auth/login.go

```diff
@@ -0,0 +1,162 @@
+
+type Authenticator interface {
+	RequestCode(context.Context) (*auth.DeviceCode, *atlas.Response, error)
+	PollToken(context.Context, *auth.DeviceCode) (*auth.Token, *atlas.Response, error)
+}
+
+type loginOpts struct {
+	OutWriter      io.Writer
+	AuthToken      string
+	RefreshToken   string
+	OpsManagerURL  string
+	ProjectID      string
+	OrgID          string
+	MongoShellPath string
+	isGov          bool
+	isCloudManager bool
+	noBrowser      bool
+	config         config.SetSaver
+	flow           Authenticator
+}
+
+func (opts *loginOpts) initFlow() error {
+	var err error
+	opts.flow, err = oauth.FlowWithConfig(config.Default())
+	return err
+}
+
+func (opts *loginOpts) SetUpAccess() {
+	if opts.isGov {
+		opts.config.Set("service", config.CloudGovService)
+	} else if opts.isCloudManager {
+		opts.config.Set("service", config.CloudManagerService)
+	}
+
+	if opts.AuthToken != "" {
+		opts.config.Set("auth_token", opts.AuthToken)
+	}
+	if opts.RefreshToken != "" {
+		opts.config.Set("refresh_token", opts.RefreshToken)
+	}
+	if opts.OpsManagerURL != "" {
+		opts.config.Set("ops_manager_url", opts.OpsManagerURL)
+	}
+}
+
+func (opts *loginOpts) Run(ctx context.Context) error {
+	code, _, err := opts.flow.RequestCode(ctx)
+	if err != nil {
+		return err
+	}
+
+	codeDuration := time.Duration(code.ExpiresIn) * time.Second
+	_, _ = fmt.Fprintf(opts.OutWriter, `
+Here is your one-time code: %s-%s
+
+Sign in with your browser and enter the code.
+
+Or go to %s
+
+Your code will expire after %.0f minutes.
+`,
+		code.UserCode[0:len(code.UserCode)/2],
+		code.UserCode[len(code.UserCode)/2:],
+		code.VerificationURI,
+		codeDuration.Minutes(),
+	)
+	if !opts.noBrowser {
+		if errBrowser := browser.OpenURL(code.VerificationURI); errBrowser != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "There was an issue opening your browser\n")
+		}
+	}
+
+	accessToken, _, err := opts.flow.PollToken(ctx, code)
+	if err != nil {
+		return err
+	}
+	opts.AuthToken = accessToken.AccessToken
+	opts.RefreshToken = accessToken.RefreshToken
+	opts.SetUpAccess()
+	if err := opts.config.Save(); err != nil {
+		return err
+	}
+	_, _ = fmt.Fprintf(opts.OutWriter, "Successfully logged in\n")
+	return nil
+}
+
+func LoginBuilder() *cobra.Command {
+	opts := &loginOpts{}
+	cmd := &cobra.Command{
+		Use:   "login",
+		Short: "Authenticate with Atlas",
+		Example: `  To start the interactive setup
+  $ mongocli auth login
+`,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			opts.OutWriter = cmd.OutOrStdout()
+			opts.config = config.Default()
+			return opts.initFlow()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return opts.Run(cmd.Context())
+		},
+		Annotations: map[string]string{
+			"toc": "true",
+		},
+		Args: require.NoArgs,
+	}
+
+	cmd.Flags().BoolVar(&opts.isGov, "gov", false, "Loging to Atlas for Government.")
+	cmd.Flags().BoolVar(&opts.isCloudManager, "cm", false, "Loging to Cloud Manager.")
```

Comment:
Appears to be a typo here?

```suggestion
	cmd.Flags().BoolVar(&opts.isGov, "gov", false, "Log in to Atlas for Government.")
	cmd.Flags().BoolVar(&opts.isCloudManager, "cm", false, "Log in to Cloud Manager.")
```

Reply:
oops long day